File: algorithms/code/leetcode/lc020_valid_parentheses/lc020_valid_parentheses.py
# https://leetcode.com/problems/valid-parentheses/

import unittest


class Solution:
    def check(self, s: str) -> bool:
        return len(s) == 0

    def isValid(self, s: str) -> bool:
        result = self.check(s)
        if result:
            return True
        stack = []
        lut = {')': '(', '}': '{', ']': '['}
        for character in s:
            if character in lut.values():
                stack.append(character)
            elif not stack or stack.pop() != lut[character]:
                return False
        return len(stack) == 0


class TestCode(unittest.TestCase):
    def test_garbage(self) -> None:
        assert not Solution().isValid('*')

    def test_RB(self) -> None:
        assert not Solution().isValid('(')

    def test_RE(self) -> None:
        assert not Solution().isValid(')')

    def test_RB_RE(self) -> None:
        assert Solution().isValid('()')

    def test_RB_RE_SB_SE_CB_CE(self) -> None:
        assert Solution().isValid('()[]{}')

    def test_RB_SB_RE_SE(self) -> None:
        assert not Solution().isValid('([)]')

    def test_RB_CB_RE_CE(self) -> None:
        assert not Solution().isValid('({)}')

    def test_SB_RB_SE_RE(self) -> None:
        assert not Solution().isValid('[(])')

    def test_CB_RB_CE_RE(self) -> None:
        assert not Solution().isValid('{(})')

    def test_RE__SE__CE(self) -> None:
        assert not Solution().isValid(')')
        assert not Solution().isValid(']')
        assert not Solution().isValid('}')

    def test_nothing(self) -> None:
        assert Solution().isValid('')

File: algorithms/code/leetcode/lc020_valid_parentheses/lc020_valid_parentheses.rs
// https://leetcode.com/problems/valid-parentheses/

use std::collections::HashMap;

pub struct Solution;

impl Solution {
    pub fn is_valid(s: String) -> bool {
        if s.is_empty() {
            return true;
        }
        let mut stack = vec![];
        let lut: HashMap<char, char> = [(')', '('), ('}', '{'), (']', '[')]
            .iter()
            .cloned()
            .collect();
        for character in s.chars() {
            if lut.values().any(|&value| value == character) {
                stack.push(character);
            } else if stack.is_empty() || *lut.get(&character).unwrap() != stack.pop().unwrap() {
                return false;
            }
        }
        stack.is_empty()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_garbage() {
        assert!(!Solution::is_valid("*".to_string()));
    }

    #[test]
    fn test_rb() {
        assert!(!Solution::is_valid("(".to_string()));
    }

    #[test]
    fn test_re() {
        assert!(!Solution::is_valid(")".to_string()));
    }

    #[test]
    fn test_rb_re() {
        assert!(Solution::is_valid("()".to_string()));
    }

    #[test]
    fn test_rb_re_sb_se_cb_ce() {
        assert!(Solution::is_valid("()[]{}".to_string()));
    }

    #[test]
    fn test_rb_sb_re_se() {
        assert!(!Solution::is_valid("([)]".to_string()));
    }

    #[test]
    fn test_rb_cb_re_ce() {
        assert!(!Solution::is_valid("({)}".to_string()));
    }

    #[test]
    fn test_sb_rb_se_re() {
        assert!(!Solution::is_valid("[(])".to_string()));
    }

    #[test]
    fn test_cb_rb_ce_re() {
        assert!(!Solution::is_valid("{(})".to_string()));
    }

    #[test]
    fn test_re_se_ce() {
        assert!(!Solution::is_valid(")".to_string()));
        assert!(!Solution::is_valid("]".to_string()));
        assert!(!Solution::is_valid("}".to_string()));
    }

    #[test]
    fn test_nothing() {
        assert!(Solution::is_valid("".to_string()));
    }
}
